packages/astro/src/core/build/internal.ts

```diff
@@ -240,39 +240,54 @@ export function* eachServerPageData(internals: BuildInternals) {
 	}
 }
 
+interface OrderInfo {
+	depth: number;
+	order: number;
+}
+
 /**
  * Sort a page's CSS by depth. A higher depth means that the CSS comes from shared subcomponents.
  * A lower depth means it comes directly from the top-level page.
  * The return of this function is an array of CSS paths, with shared CSS on top
  * and page-level CSS on bottom.
  */
-export function sortedStylesheets(pageData: PageBuildData) {
-	return pageData.styles
-		.sort((a, b) => {
-			let depthA = a.depth,
-				depthB = b.depth,
-				orderA = a.order,
-				orderB = b.order;
-
-			if (orderA === -1 && orderB >= 0) {
-				return 1;
-			} else if (orderB === -1 && orderA >= 0) {
-				return -1;
-			} else if (orderA > orderB) {
-				return 1;
-			} else if (orderA < orderB) {
-				return -1;
-			} else {
-				if (depthA === -1) {
-					return -1;
-				} else if (depthB === -1) {
-					return 1;
-				} else {
-					return depthA > depthB ? -1 : 1;
-				}
-			}
-		})
-		.map(({ sheet }) => sheet);
+export function cssOrder(a: OrderInfo, b: OrderInfo) {
+	let depthA = a.depth,
+		depthB = b.depth,
+		orderA = a.order,
+		orderB = b.order;
+
+	if (orderA === -1 && orderB >= 0) {
+		return 1;
+	} else if (orderB === -1 && orderA >= 0) {
+		return -1;
+	} else if (orderA > orderB) {
+		return 1;
+	} else if (orderA < orderB) {
+		return -1;
+	} else {
+		if (depthA === -1) {
+			return -1;
+		} else if (depthB === -1) {
+			return 1;
+		} else {
+			return depthA > depthB ? -1 : 1;
+		}
+	}
+}
+
+export function mergeInlineCss(
```

Comment:
perf nit (premature optimization + I'm not familiar with Astro)

I'd guess that for sufficiently large projects with a non-trivial amount of CSS, the excess copying won't be trivial.
A regular `for` loop and pushing merged/non-inline `StyleSheetAsset`s to an array avoids the issue.